sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -313,14 +318,57 @@ def lloyd(
         # refers. For this reason, this strategy is not compatible with sklearn
         # unit tests, that consider new_centroids_t (the array after the update)
         # to be the best centroids at each iteration.
-
         centroids_t, new_centroids_t = (new_centroids_t, centroids_t)
 
+        # ???: if two successive assignations have been computed equal, it's called
+        # "strict convergence" and means that the algorithm has converged and can't get
+        # better (since same assignments will produce the same centroid updates, which
+        # will produce the same assignments, and so on..). scikit-learn decides to
+        # check for strict convergence at each iteration, but that sounds expensive
+        # since it requires an additional pass on each sample label every time.
+        # Providing the user chooses a sensible value for `tol`, wouldn't the cost of
+        # this check be in general greater than what the benefits ?
+
+        # Following this reasoning, unlike scikit-learn, we choose to enforce this
+        # behavior if and only if `tol == 0`, because in this case, it is easy to see
+        # that lloyd can indeed fail to stop at the right time due to numerical errors.
+        # Moreover, this is enough to pass scikit-learn unit tests. When `tol > 0`, we
+        # rely on the user setting an appropriate tolerance threshold.
```

Comment:
I would rather track the exact behavior of scikit-learn w.r.t. convergence checks.